## 4.8 Grouping Data & Aggregating Variables - Task 4.8

## Contents
#### 1. Task 4.8

## Task 4.8

#### Question 1 - Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#Create path
path = r'/Users/anfran/Documents/Instacart Basket Analysis'

In [3]:
#Import data set
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_updated_4.8.pkl'))

In [4]:
ords_prods_merge.shape

(32404859, 19)

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0               9999999.0       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  Regularly busy  Regular days        Average orders         10  New customer  
1  Regularly busy  Slowest days        Average orders         10  New customer  
2  Regularly busy  Slowest days           Most orders         10  New customer  
3      Least busy  Slowest days        Average orders         10  New customer  
4      Least busy  Slowest days           Most orders         10  New customer

#### Question 2 - In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [7]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### Question 3 - Analyze the result. How do the results for the entire dataframe differ from those of the subset? Include your comments in a markdown cell below the executed code.

The results from the entire dataframe show more departments than from the subset, allowing the analyst to perform a more complete analysis. The averages are different also. For example. in the subset, the average for department_id 4 was 18.825780. When considering the entire dataframe, the average for department_id 4 was 17.811403. I would conclude that the maverage from the entire dataframe gives the analyst a more complete picture. 

#### Question 4 - Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [8]:
#This was already accomplished. Please see the 4.8 Grouping Data & Aggregating Variables - Exercise 4.8 notebook.
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0               9999999.0       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  Regularly busy  Regular days        Average orders         10  New customer  
1  Regularly busy  Slowest days        Average orders         10  New customer  
2  Regularly busy  Slowest days           Most orders         10  New customer  
3      Least busy  Slowest days        Average orders         10  New customer  
4      Least busy  Slowest days           Most orders         10  New customer

#### Question 5 - The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [9]:
#Check statistics for each 'loyalty_flag' category
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

New customers's product prices have the highest average. Regular customers have the highest total sales. 

#### Question 6 - The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
#### 1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### 2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [20]:
#Create 'average_price' Column
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [22]:
#Round averages
ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

In [23]:
#Check results
ords_prods_merge.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                9999999.0       196.0                1.0        0.0   
1                     15.0       196.0                1.0        1.0   
2                     21.0       196.0                1.0        1.0   
3                     29.0       196.0                1.0        1.0   
4                     28.0       196.0                1.0        1.0   
5                     19.0       196.0                1.0        1.0   
6                     20.0       196.0                1.0        1.0   
7                     14.0       196.0                2.0        1.0   
8                      0.0       196.0                4.0        1.0   
9                     30.0       196.0                1.0        1.0   
10                     7.0       196.0                2.0        0.0   
11                     8.0       196.0                6.0        1.0   
12                     7.0       196.0                1.0        1.0   
13                     7.0       196.0                1.0        1.0   
14                    14.0       196.0                1.0        1.0   
15                     6.0       196.0                3.0        0.0   
16                     8.0       196.0               15.0        1.0   
17                     6.0       196.0                1.0        1.0   
18                     6.0       196.0                2.0        0.0   
19                     7.0       196.0               17.0        0.0   
20                     8.0       196.0               14.0        1.0   
21                     4.0       196.0                9.0        0.0   
22                     3.0       196.0                1.0        1.0   
23               9999999.0       196.0                4.0        0.0   
24                     7.0       196.0                5.0        1.0   

   product_name  ...  department_id  prices    price_range_loc  \
0          Soda  ...              7 

In [24]:
#Create spending_flag column
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'Spending_flag'] = 'High spender'

In [25]:
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'Spending_flag'] = 'Low spender'

In [26]:
#Check results
ords_prods_merge.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                9999999.0       196.0                1.0        0.0   
1                     15.0       196.0                1.0        1.0   
2                     21.0       196.0                1.0        1.0   
3                     29.0       196.0                1.0        1.0   
4                     28.0       196.0                1.0        1.0   
5                     19.0       196.0                1.0        1.0   
6                     20.0       196.0                1.0        1.0   
7                     14.0       196.0                2.0        1.0   
8                      0.0       196.0                4.0        1.0   
9                     30.0       196.0                1.0        1.0   
10                     7.0       196.0                2.0        0.0   
11                     8.0       196.0                6.0        1.0   
12                     7.0       196.0                1.0        1.0   
13                     7.0       196.0                1.0        1.0   
14                    14.0       196.0                1.0        1.0   
15                     6.0       196.0                3.0        0.0   
16                     8.0       196.0               15.0        1.0   
17                     6.0       196.0                1.0        1.0   
18                     6.0       196.0                2.0        0.0   
19                     7.0       196.0               17.0        0.0   
20                     8.0       196.0               14.0        1.0   
21                     4.0       196.0                9.0        0.0   
22                     3.0       196.0                1.0        1.0   
23               9999999.0       196.0                4.0        0.0   
24                     7.0       196.0                5.0        1.0   

   product_name  ...  department_id  prices    price_range_loc  \
0          Soda  ...              7 

In [27]:
ords_prods_merge.shape

(32404859, 21)

#### Question 7 - In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
#### 1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
#### 2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
#### 3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [28]:
#Create 'median_days_since_prior_order' column
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [30]:
#Check results
ords_prods_merge.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                9999999.0       196.0                1.0        0.0   
1                     15.0       196.0                1.0        1.0   
2                     21.0       196.0                1.0        1.0   
3                     29.0       196.0                1.0        1.0   
4                     28.0       196.0                1.0        1.0   
5                     19.0       196.0                1.0        1.0   
6                     20.0       196.0                1.0        1.0   
7                     14.0       196.0                2.0        1.0   
8                      0.0       196.0                4.0        1.0   
9                     30.0       196.0                1.0        1.0   
10                     7.0       196.0                2.0        0.0   
11                     8.0       196.0                6.0        1.0   
12                     7.0       196.0                1.0        1.0   
13                     7.0       196.0                1.0        1.0   
14                    14.0       196.0                1.0        1.0   
15                     6.0       196.0                3.0        0.0   
16                     8.0       196.0               15.0        1.0   
17                     6.0       196.0                1.0        1.0   
18                     6.0       196.0                2.0        0.0   
19                     7.0       196.0               17.0        0.0   
20                     8.0       196.0               14.0        1.0   
21                     4.0       196.0                9.0        0.0   
22                     3.0       196.0                1.0        1.0   
23               9999999.0       196.0                4.0        0.0   
24                     7.0       196.0                5.0        1.0   

   product_name  ...  prices    price_range_loc     busiest_day  busiest_days  \
0          Soda  ... 

In [31]:
#Create 'order_frequency_flag' column
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [32]:
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'order_frequency_flag'] = 'Regular customer'

In [33]:
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [34]:
#Check results
ords_prods_merge.head(25)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                9999999.0       196.0                1.0        0.0   
1                     15.0       196.0                1.0        1.0   
2                     21.0       196.0                1.0        1.0   
3                     29.0       196.0                1.0        1.0   
4                     28.0       196.0                1.0        1.0   
5                     19.0       196.0                1.0        1.0   
6                     20.0       196.0                1.0        1.0   
7                     14.0       196.0                2.0        1.0   
8                      0.0       196.0                4.0        1.0   
9                     30.0       196.0                1.0        1.0   
10                     7.0       196.0                2.0        0.0   
11                     8.0       196.0                6.0        1.0   
12                     7.0       196.0                1.0        1.0   
13                     7.0       196.0                1.0        1.0   
14                    14.0       196.0                1.0        1.0   
15                     6.0       196.0                3.0        0.0   
16                     8.0       196.0               15.0        1.0   
17                     6.0       196.0                1.0        1.0   
18                     6.0       196.0                2.0        0.0   
19                     7.0       196.0               17.0        0.0   
20                     8.0       196.0               14.0        1.0   
21                     4.0       196.0                9.0        0.0   
22                     3.0       196.0                1.0        1.0   
23               9999999.0       196.0                4.0        0.0   
24                     7.0       196.0                5.0        1.0   

   product_name  ...    price_range_loc     busiest_day  busiest_days  \
0          Soda  ...  Mid-ran

In [35]:
ords_prods_merge.shape

(32404859, 23)

In [37]:
ords_prods_merge['order_frequency_flag'].value_counts()

Frequent customer        20514478
Regular customer          7269731
Non-frequent customer     4620650
Name: order_frequency_flag, dtype: int64

#### Question 9 - Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [38]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined_updated_4.8.2.pkl'))